### DISCUSSION TITLE
Multiple types of o/auth types at the same time

### DISCUSSION THREAD

**Discussion by maverich** [maxcountryman/axum-login#147] [Q&A]
Hello, I am trying to build something with multiple o/auth types. By advice from Max, I've used the enum type for credentials like as follows:

```
#[derive(Debug, Clone, Deserialize)]
pub enum Credentials{
    BasicCred(BasicCred)
}

// Stands for 
#[derive(Debug, Clone, Deserialize)]
pub struct BasicCred {
    pub username: String,
    pub password: String,
    pub next: Option<String>,
}
```

And my `authenticate` simply looks like this:

```
#[async_trait]
impl AuthnBackend for Backend {
    type User = User;
    type Credentials = Credentials;
    type Error = sqlx::Error;

    async fn authenticate(
        &self,
        creds: Self::Credentials,
    ) -> Result<Option<Self::User>, Self::Error> {
        match creds {
            Self::Credentials::BasicCred(upcred) =>{
                // Check with the DB and authenticate or authorise ...
            }
        }
    }
}
```
When I try to do the authenticate using the basic credentials on the frontend side I receive 422 with following error:
`Failed to deserialize form body: invalid type: map, expected enum Credentials`.
Any help appreciated :)

**Comment by maxcountryman** [comment]
> I am trying to build something with multiple o/auth types

I'm reading this as multiple OAuth providers, like GitHub and Google, as examples. Is that right?

What I would suggest for this then is a variant per provider. Then you'd need separate routes for each that would explicitly map to that variant. Please note, you'll need a separate type for the form serialization to work.

We would probably benefit from having an example that demonstrates this.

**Reply by maverich** [reply]
The moment I resolve this, I can write an example that is like combination of basic auth and oauth with Github, share it here and as a PR to examples directory happily :) 

I've tried separating the `login` routes though that created some other problems with permission_required! macro. Just for clarification; I've referenced a basic_login endpoint for one and git_login for the other one.

**Reply by maxcountryman** [reply]
Are you able to share what you have so far? Happy to take a look if so.

**Reply by maverich** [reply]
I've solved it :) I will write the example tomorrow, for now it is in a private repo.
